File: eggs-gui/web/daemon_client.py
"""JSON-RPC client for the eggs-gui daemon over Unix socket."""

import json
import socket
import threading
from typing import Any, Callable, Optional

SOCKET_PATH = "/tmp/eggs-gui.sock"


class DaemonClient:
    """Connects to the eggs-gui Go daemon via JSON-RPC over Unix socket."""

    def __init__(self):
        self._sock: Optional[socket.socket] = None
        self._next_id = 0
        self._lock = threading.Lock()

    def connect(self):
        self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self._sock.connect(SOCKET_PATH)

    def close(self):
        if self._sock:
            self._sock.close()
            self._sock = None

    def call(self, method: str, params: Any = None) -> Any:
        """Make a JSON-RPC call and return the result."""
        with self._lock:
            self._next_id += 1
            request = {
                "jsonrpc": "2.0",
                "method": method,
                "params": params or {},
                "id": self._next_id,
            }
            self._send(request)

            while True:
                resp = self._recv()
                # Skip notifications
                if "id" not in resp or resp.get("id") is None:
                    continue
                if resp.get("error"):
                    raise Exception(resp["error"]["message"])
                return resp.get("result")

    def call_stream(
        self,
        method: str,
        params: Any = None,
        on_line: Optional[Callable[[str], None]] = None,
    ) -> Any:
        """Make a JSON-RPC call that streams output via notifications."""
        with self._lock:
            self._next_id += 1
            request = {
                "jsonrpc": "2.0",
                "method": method,
                "params": params or {},
                "id": self._next_id,
            }
            self._send(request)

            while True:
                resp = self._recv()
                # Handle notification
                if ("id" not in resp or resp.get("id") is None) and resp.get("method") == "stream.output":
                    line = resp.get("params", {}).get("line", "")
                    if on_line:
                        on_line(line)
                    continue
                # Final response
                if resp.get("error"):
                    raise Exception(resp["error"]["message"])
                return resp.get("result")

    def _send(self, data: dict):
        raw = json.dumps(data) + "\n"
        self._sock.sendall(raw.encode())

    def _recv(self) -> dict:
        buf = b""
        while True:
            chunk = self._sock.recv(4096)
            if not chunk:
                raise ConnectionError("Daemon disconnected")
            buf += chunk
            if b"\n" in buf:
                line, _ = buf.split(b"\n", 1)
                return json.loads(line)

File: eggs-ai/src/bridge/daemon-client.ts
import { connect } from 'node:net';

/**
 * JSON-RPC client for the eggs-gui daemon.
 * Connects over Unix socket at /tmp/eggs-gui.sock.
 *
 * This lets eggs-ai read live system state from the daemon
 * instead of re-inspecting the system directly — giving it
 * the same data the GUI frontends see.
 */

const DEFAULT_SOCKET = '/tmp/eggs-gui.sock';

interface JsonRpcRequest {
  jsonrpc: '2.0';
  id: number;
  method: string;
  params?: Record<string, unknown>;
}

interface JsonRpcResponse {
  jsonrpc: '2.0';
  id: number;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
}

export class DaemonClient {
  private socketPath: string;
  private nextId = 1;

  constructor(socketPath = DEFAULT_SOCKET) {
    this.socketPath = socketPath;
  }

  /**
   * Send a JSON-RPC request to the daemon and return the result.
   */
  async call<T = unknown>(method: string, params?: Record<string, unknown>): Promise<T> {
    return new Promise((resolve, reject) => {
      const request: JsonRpcRequest = {
        jsonrpc: '2.0',
        id: this.nextId++,
        method,
        params,
      };

      const socket = connect(this.socketPath);
      let data = '';

      socket.on('connect', () => {
        socket.write(JSON.stringify(request) + '\n');
      });

      socket.on('data', (chunk) => {
        data += chunk.toString();
        // Try to parse complete JSON-RPC responses (newline-delimited)
        const lines = data.split('\n');
        for (const line of lines) {
          if (!line.trim()) continue;
          try {
            const response = JSON.parse(line) as JsonRpcResponse;
            if (response.id === request.id) {
              socket.end();
              if (response.error) {
                reject(new Error(`Daemon error: ${response.error.message}`));
              } else {
                resolve(response.result as T);
              }
            }
          } catch {
            // Incomplete JSON, wait for more data
          }
        }
      });

      socket.on('error', (err) => {
        reject(new Error(`Cannot connect to eggs-gui daemon at ${this.socketPath}: ${err.message}`));
      });

      socket.setTimeout(30000, () => {
        socket.destroy();
        reject(new Error('Daemon request timed out'));
      });
    });
  }

  /**
   * Check if the daemon is running and reachable.
   */
  async isAvailable(): Promise<boolean> {
    try {
      await this.call('system.versions');
      return true;
    } catch {
      return false;
    }
  }

  // ─── Typed convenience methods matching eggs-gui.json schema ───

  async getConfig(): Promise<Record<string, unknown>> {
    return this.call('config.read');
  }

  async getVersions(): Promise<{ eggs: string; calamares: string; distro: string; kernel: string }> {
    return this.call('system.versions');
  }

  async checkDeps(): Promise<{ eggsInstalled: boolean; calamaresInstalled: boolean; configExists: boolean }> {
    return this.call('system.checkDeps');
  }

  async listISOs(): Promise<Array<{ path: string; name: string; size: number; modified: string }>> {
    return this.call('iso.list');
  }

  async listWardrobe(): Promise<{ costumes: string[]; accessories: string[]; servers: string[] }> {
    return this.call('wardrobe.list');
  }

  async getStatus(): Promise<{ taskId: string }> {
    return this.call('eggs.status');
  }
}
